Main arguments checker

gle-mini opened the pull request:
Create a new class `Password` and his method is_valid.
Create a function check_main_arguments used in the Makefile.

gle-mini committed:
create Password class [DOING]

gle-mini committed:
Merge branch 'main' into main-arguments-checker

gle-mini committed:
main arguments checker [OK]

gle-mini committed:
Restores alphabetical ordering of files in the Makefile

gle-mini committed:
add header comment and define in

Saphirac commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+
+	size_t const length = this->length();
+
+	char const firstChar = this->at(0);
```
Please use snake case for variable definition

Saphirac commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+	return (c >= 0x01 && c <= 0x09) || (c >= 0x0B && c <= 0x0C) || (c >= 0x0E && c <= 0x1F) || (c >= 0x21);
+}
+
+static bool is_nolfcr(char c) { return c != '\n' && c != '\r'; }
```
these 3 functions can be inline static

Saphirac commented on src/main.cpp:
```diff
@@ -3,28 +3,68 @@
+/**
+ * @brief check if the main arguments are correct
+ *
+ * The server can only use ports with numbers 1024-49151 that are registered ports;
```
not exactly that

Saphirac commented on src/main.cpp:
```diff
@@ -3,28 +3,68 @@
+		std::cerr << "incorrect: number of arguments" << std::endl;
+		return false;
+	}
+	for (const char *p = av[1]; *p; ++p)
```
why p ?

Saphirac commented on src/main.cpp:
```diff
@@ -3,28 +3,68 @@
+			return false;
+		}
+	}
+	int const port = atoi(av[1]);
```
what about port > INT_MAX ?

Saphirac reviewed (changes requested):
Please change these few things

gle-mini committed:
Fix pull request issues
change variable name and change the functions to inline static `Password` class [OK]
change the header comment and secure the case port > INT_MAX in
`main.cpp` [OK]

Saphirac commented on src/main.cpp:
```diff
@@ -3,28 +3,81 @@
 	}
 
+	char *end_ptr;
+	errno = 0;
```
norm error ?

Saphirac reviewed (commented):
Norm error i think

gle-mini committed:
Fix norm error [OK]

Saphirac reviewed (dismissed):
Let's merge that :D

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/core.cpp:
```diff
@@ -0,0 +1,24 @@
+// Constructors //
+
+/**
+ * @param password The nickname to initialize the instance with.
```
Incorrect doxygen comment. Please, change it.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+inline static bool is_nolfcrspcl(char c)
+{
+	return (c >= 0x01 && c <= 0x09) || (c >= 0x0B && c <= 0x0C) || (c >= 0x0E && c <= 0x1F) || (c >= 0x21 && c <= 0x39)
+	    || (c >= 0x3B);
```
This way is C style, you'd better do like this:
```c++
return std::string("\r\n :").find(c) == std::string::npos;
```

gle-mini replied:
I corrected it like that do you approve ?



```c++
inline static bool is_nolfcrspcl(char c)
{
	return std::string("\x00\r\n :").find(c) == std::string::npos;
}
```

JonathanDUFOUR replied:
No, you don't need to look for the `NUL` character. Indeed, the `std::string::find` method doesn't allow you to match `NUL` characters. If you try to do the following:
```c++
size_t const nul_pos = std::string("You won't get it").find('\0');
```
then you will ALWAYS get `std::string::npos` instead of the actual position of the `NUL` character.

JonathanDUFOUR replied:
Even doing the following:
```c++
size_t const nul_pos = std::string("Neither\0this way").find('\0');
```

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+
+inline static bool is_nolfcrsp(char c)
+{
+	return (c >= 0x01 && c <= 0x09) || (c >= 0x0B && c <= 0x0C) || (c >= 0x0E && c <= 0x1F) || (c >= 0x21);
```
This way is C style, you'd better do like this:
```c++
return std::string("\r\n ").find(c) == std::string::npos;
```

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+
+#define MAXIMUM_LENGTH_FOR_PASSWORD 50
+
+inline static bool is_nolfcrspcl(char c)
```
Missing `const` qualifier for `c`.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+	    || (c >= 0x3B);
+}
+
+inline static bool is_nolfcrsp(char c)
```
Missing `const` qualifier for `c`.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+	return (c >= 0x01 && c <= 0x09) || (c >= 0x0B && c <= 0x0C) || (c >= 0x0E && c <= 0x1F) || (c >= 0x21);
+}
+
+inline static bool is_nolfcr(char c) { return c != '\n' && c != '\r'; }
```
Missing `const` qualifier for `c` + This way is C style, you'd better do like this:
```c++
return std::string("\r\n").find(c) == std::string::npos;
```

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+
+	size_t const length = this->length();
+
+	char const first_char = this->at(0);
```
Norm error + it is unnecessary to have a constant here for a unique usage. The only exception to this is when it's more explicit to do so, but here, we all know that `this->at(0)` is the string's first character.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+
+	for (size_t i = 1; i < length; ++i)
+	{
+		char c = this->at(i);
```
Norm error.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+			}
+			break; // Exit the loop after processing the portion following ':'
+		}
+		else if (!is_nolfcrsp(c))
```
Unnecessary `else` statement, you `break` if you fall in the previous `if` statement, so you can only reach what is after it if the condition was false.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,64 @@
+				if (!is_nolfcr(this->at(j)))
+					return false;
+			}
+			break; // Exit the loop after processing the portion following ':'
```
Here you'd better directly `return true` instead of `break`, which is a `goto` to the end of the `for` loop, and then a `return true`.

JonathanDUFOUR commented on src/main.cpp:
```diff
@@ -3,27 +3,81 @@
+		std::cerr << "incorrect: number of arguments" << std::endl;
+		return false;
 	}
+	for (const char *it = av[1]; *it; ++it)
```
Please, put the `const` qualifier at the right of the type that it qualifies + `it` is quite unadapted because it's not an iterator, but a raw pointer. You may call it `ptr` or `current_char` instead. Or even if you want to keep this name, you should name it `cit` instead because it's more like a `const_iterator`.

JonathanDUFOUR commented on src/main.cpp:
```diff
@@ -3,27 +3,81 @@
+
+	errno = 0;
+
+	if ((errno == ERANGE && (port == LONG_MAX || port == LONG_MIN)) || (errno != 0 && port == 0))
```
Only checking that `errno == ERANGE` is enough, there's no need to check more.

JonathanDUFOUR commented on src/main.cpp:
```diff
@@ -3,27 +3,81 @@
+		return false;
+	}
+
+	if (*end_ptr != '\0' || !(port >= MINIMUM_PORT && port <= MAXIMUM_PORT))
```
Checking that `port < MINIMUM_PORT || port > MAXIMUM_PORT` would compute 1 less instruction.

JonathanDUFOUR commented on src/main.cpp:
```diff
@@ -3,27 +3,81 @@
+
+	if (*end_ptr != '\0' || !(port >= MINIMUM_PORT && port <= MAXIMUM_PORT))
+	{
+		std::cerr << "incorrect: port must be between 1024 and 49151" << std::endl;
```
You should use the MACROS that you defined before instead of hardcoding the values in the error string.

JonathanDUFOUR reviewed (changes requested):
Good job, but some details remain to be changed. Please read my comments :)

gle-mini committed:
Corrected pull request change [OK]

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,65 @@
+
+inline static bool is_nolfcrspcl(char const c)
+{
+	return std::string("\x00\r\n :").find(c) == std::string::npos;
```
As I replied in the thread of my previous comment, you won't ever be able to match `NUL` characters using the `std::string::find` method, so please, remove it from the characters to match.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,65 @@
+
+inline static bool is_nolfcrsp(char const c)
+{
+	return std::string("\x00\r\n ").find(c) == std::string::npos;
```
As I replied in the thread of my previous comment, you won't ever be able to match NUL characters using the std::string::find method, so please, remove it from the characters to match.

JonathanDUFOUR commented on src/lib/class/specialized_string/Password/methods.cpp:
```diff
@@ -0,0 +1,65 @@
+
+inline static bool is_nolfcr(char const c)
+{
+	return std::string("\x00\r\n").find(c) == std::string::npos;
```
As I replied in the thread of my previous comment, you won't ever be able to match NUL characters using the std::string::find method, so please, remove it from the characters to match.

JonathanDUFOUR commented on src/main.cpp:
```diff
@@ -3,27 +3,81 @@
+#include "class/specialized_string/Password.hpp"
+#include <cctype>
+#include <cerrno>
+#include <climits>
```
You don't need `#include <climits>` anymore, please remove it.

JonathanDUFOUR reviewed (changes requested):
Please change the last things I reported and then we are good! :D

gle-mini committed:
Corrected pull request change [OK]

JonathanDUFOUR reviewed (approved):
Great work! Everything is good now! Keep it up!
